File: backend/app/services/numbering_generator.py
# ============================================================================
# FILE: backend/app/services/numbering_generator.py
# ============================================================================

from typing import Dict, Any, Optional
import logging

logger = logging.getLogger(__name__)


class NumberingGenerator:
    """Generate numbering sequences for tickets"""
    
    def __init__(self, numbering_config: Dict[str, Any]):
        self.config = numbering_config
    
    def generate_for_ticket(self, ticket_number: int, row_data: Optional[Dict[str, Any]] = None) -> str:
        """
        Generate numbering string for a specific ticket
        
        Args:
            ticket_number: Ticket index (1-based)
            row_data: Optional data from CSV/XLSX import
        
        Returns:
            Generated numbering string
        """
        elements = self.config.get('elements', [])
        parts = []
        
        for element in sorted(elements, key=lambda x: x.get('order', 0)):
            element_parts = []
            
            for sub_element in sorted(element.get('subElements', []), key=lambda x: x.get('order', 0)):
                sub_type = sub_element.get('type')
                config = sub_element.get('config', {})
                
                if sub_type == 'text':
                    # Static text
                    text = config.get('text', '')
                    element_parts.append(text)
                
                elif sub_type == 'numeric':
                    # Dynamic numeric value
                    value = self._generate_numeric(config, ticket_number)
                    element_parts.append(value)
                
                # Check if column mapping exists (CSV/XLSX import)
                column_name = sub_element.get('columnMapping')
                if column_name and row_data:
                    # Override with data from import
                    element_parts[-1] = str(row_data.get(column_name, element_parts[-1]))
            
            # Join sub-elements with separator
            separator_config = element.get('separator', {})
            separator = self._get_separator(separator_config)
            element_text = separator.join(element_parts)
            
            parts.append(element_text)
        
        return ''.join(parts)
    
    def _generate_numeric(self, config: Dict[str, Any], ticket_number: int) -> str:
        """Generate numeric value based on configuration"""
        numeric_type = config.get('numericType', '123')
        start_value = config.get('startValue', 1)
        step = config.get('step', 1)
        flow = config.get('flow', 'increment')
        
        # Calculate value
        if flow == 'increment':
            value = start_value + (ticket_number - 1) * step
        else:  # decrement
            value = start_value - (ticket_number - 1) * step
        
        # Format value
        if numeric_type == '123':
            digits = config.get('fixedDigitCount', 3) if config.get('digits') == 'fixed' else len(str(value))
            return str(value).zfill(digits)
        
        elif numeric_type == 'roman_lower':
            return self._to_roman(value).lower()
        
        elif numeric_type == 'roman_upper':
            return self._to_roman(value)
        
        elif numeric_type == 'abc_lower':
            return chr(96 + value) if value <= 26 else f"z{value - 26}"
        
        elif numeric_type == 'abc_upper':
            return chr(64 + value) if value <= 26 else f"Z{value - 26}"
        
        elif numeric_type == 'hex':
            return hex(value)[2:].upper()
        
        return str(value)
    
    def _to_roman(self, num: int) -> str:
        """Convert integer to Roman numeral"""
        val = [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1]
        syms = ["M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"]
        roman_num = ''
        i = 0
        while num > 0:
            for _ in range(num // val[i]):
                roman_num += syms[i]
                num -= val[i]
            i += 1
        return roman_num
    
    def _get_separator(self, separator_config: Dict[str, Any]) -> str:
        """Get separator character"""
        sep_type = separator_config.get('type', 'none')
        
        separators = {
            'hyphen': '-',
            'underscore': '_',
            'dot': '.',
            'comma': ',',
            'space': ' ',
            'none': '',
            'custom': separator_config.get('customValue', '')
        }
        
        return separators.get(sep_type, '')

File: frontend/lib/ticket-layout/numberingGenerator.ts
// lib/ticket-layout/numberingGenerator.ts

import { NumberingElement, SubElement } from "./zustandStore";

/**
 * Client-side numbering generator for real-time preview
 * Matches Python backend logic
 */
export class NumberingGenerator {
  private elements: NumberingElement[];

  constructor(elements: NumberingElement[]) {
    this.elements = elements;
  }

  /**
   * Generate numbering text for a specific ticket
   */
  generateForTicket(ticketNumber: number, rowData?: Record<string, any>): string {
    const parts: string[] = [];

    // Sort elements by order
    const sortedElements = [...this.elements].sort((a, b) => a.zIndex - b.zIndex);

    for (const element of sortedElements) {
      const elementParts: string[] = [];

      // Sort sub-elements by order
      const sortedSubElements = [...element.subElements].sort((a, b) => a.order - b.order);

      for (const subElement of sortedSubElements) {
        let value = "";

        if (subElement.type === "text") {
          // Static text
          value = subElement.config.text || "";
        } else if (subElement.type === "numeric") {
          // Dynamic numeric
          value = this.generateNumeric(subElement, ticketNumber);
        }

        // Override with CSV/XLSX data if mapped
        if (subElement.columnMapping && rowData) {
          value = String(rowData[subElement.columnMapping] ?? value);
        }

        elementParts.push(value);
      }

      // Join with separator
      const separator = this.getSeparator(element.separator);
      const elementText = elementParts.join(separator);

      parts.push(elementText);
    }

    return parts.join("");
  }

  /**
   * Generate numeric value
   */
  private generateNumeric(subElement: SubElement, ticketNumber: number): string {
    const config = subElement.config;
    const numericType = config.numericType || "123";
    const startValue = config.startValue ?? 1;
    const step = config.step ?? 1;
    const flow = config.flow || "increment";

    // Calculate value
    let value: number;
    if (flow === "increment") {
      value = startValue + (ticketNumber - 1) * step;
    } else {
      value = startValue - (ticketNumber - 1) * step;
    }

    // Format value
    switch (numericType) {
      case "123":
        const digits = config.digits === "fixed" ? (config.fixedDigitCount ?? 3) : String(value).length;
        return String(value).padStart(digits, "0");

      case "roman_lower":
        return this.toRoman(value).toLowerCase();

      case "roman_upper":
        return this.toRoman(value);

      case "abc_lower":
        return value <= 26 ? String.fromCharCode(96 + value) : `z${value - 26}`;

      case "abc_upper":
        return value <= 26 ? String.fromCharCode(64 + value) : `Z${value - 26}`;

      case "hex":
        return value.toString(16).toUpperCase();

      default:
        return String(value);
    }
  }

  /**
   * Convert number to Roman numeral
   */
  private toRoman(num: number): string {
    const values = [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1];
    const symbols = ["M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"];

    let result = "";
    let remaining = num;

    for (let i = 0; i < values.length; i++) {
      while (remaining >= values[i]) {
        result += symbols[i];
        remaining -= values[i];
      }
    }

    return result;
  }

  /**
   * Get separator character
   */
  private getSeparator(separatorConfig: { type: string; customValue?: string }): string {
    const separators: Record<string, string> = {
      hyphen: "-",
      underscore: "_",
      dot: ".",
      comma: ",",
      space: " ",
      none: "",
      custom: separatorConfig.customValue || "",
    };

    return separators[separatorConfig.type] || "";
  }
}